File: approvalcrest-junit-jupiter/src/test/java/com/github/karsaig/approvalcrest/jupiter/RouteEquivalenceTest.java
package com.github.karsaig.approvalcrest.jupiter;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInfo;

/**
 * The stack-trace route and the TestInfo route must describe the running test identically, because
 * that description is what an approved file's path is derived from. If they disagree, adding a
 * {@code TestInfo} parameter to an existing test - an unremarkable refactor - silently orphans its
 * approved file.
 *
 * <p>Equality is asserted field by field rather than with {@code sameBeanAs} so that a failure names
 * the field that diverged.
 */
public class RouteEquivalenceTest {

    private TestInfo capturedInBeforeEach;

    @BeforeEach
    void captureTestInfo(TestInfo testInfo) {
        this.capturedInBeforeEach = testInfo;
    }

    private static void assertSameIdentity(JunitJupiterTestMeta stackRoute, Junit5InfoBasedTestMeta infoRoute) {
        assertEquals(stackRoute.testClassName(), infoRoute.testClassName(), "testClassName");
        assertEquals(stackRoute.testMethodName(), infoRoute.testMethodName(), "testMethodName");
        assertEquals(stackRoute.getTestClassPath(), infoRoute.getTestClassPath(), "testClassPath");
        assertEquals(stackRoute.getApprovedDirectory(), infoRoute.getApprovedDirectory(), "approvedDirectory");
        assertEquals(stackRoute.workingDirectory(), infoRoute.workingDirectory(), "workingDirectory");
    }

    @Test
    public void publicMethodResolvesIdenticallyByBothRoutes(TestInfo testInfo) {
        assertSameIdentity(new JunitJupiterTestMeta(), new Junit5InfoBasedTestMeta(testInfo));
    }

    /** Package-private test methods are legal in Jupiter and must behave like public ones. */
    @Test
    void packagePrivateMethodResolvesIdenticallyByBothRoutes(TestInfo testInfo) {
        assertSameIdentity(new JunitJupiterTestMeta(), new Junit5InfoBasedTestMeta(testInfo));
    }

    /**
     * A display name is presentation only. If it leaked into the identity, adding one would move the
     * approved file.
     */
    @Test
    @DisplayName("a display name that is nothing like the method name")
    void displayNameDoesNotAffectIdentity(TestInfo testInfo) {
        Junit5InfoBasedTestMeta infoRoute = new Junit5InfoBasedTestMeta(testInfo);

        assertSameIdentity(new JunitJupiterTestMeta(), infoRoute);
        assertEquals("displayNameDoesNotAffectIdentity", infoRoute.testMethodName(),
                "the method name, not the display name, identifies the test");
    }

    /**
     * TestInfo captured in a @BeforeEach describes the same test as one injected into the method, so
     * a shared setup field is a valid way to supply it.
     */
    @Test
    void testInfoCapturedInBeforeEachDescribesTheSameTest(TestInfo testInfo) {
        assertEquals(new Junit5InfoBasedTestMeta(testInfo).testMethodName(),
                new Junit5InfoBasedTestMeta(capturedInBeforeEach).testMethodName());
        assertSameIdentity(new JunitJupiterTestMeta(), new Junit5InfoBasedTestMeta(capturedInBeforeEach));
    }

    @Nested
    class NestedTests {

        @Test
        void nestedClassResolvesIdenticallyByBothRoutes(TestInfo testInfo) {
            assertSameIdentity(new JunitJupiterTestMeta(), new Junit5InfoBasedTestMeta(testInfo));
        }

        /** The nested class, not the enclosing one, identifies the test. */
        @Test
        void nestedClassIsTheTestClass(TestInfo testInfo) {
            assertEquals(NestedTests.class.getName(), new Junit5InfoBasedTestMeta(testInfo).testClassName());
            assertEquals(NestedTests.class.getName(), new JunitJupiterTestMeta().testClassName());
        }
    }
}

File: approvalcrest-junit-jupiter-kotlin/src/test/kotlin/com/github/karsaig/approvalcrest/kotlin/RouteEquivalenceTest.kt
package com.github.karsaig.approvalcrest.kotlin

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInfo

/**
 * The stack-trace route and the TestInfo route must describe the running test identically, because
 * that description determines the approved file's path. If they disagreed, adding a [TestInfo]
 * parameter to an existing test would silently orphan its approved file.
 */
class RouteEquivalenceTest {

    private var capturedInBeforeEach: TestInfo? = null

    @BeforeEach
    fun captureTestInfo(testInfo: TestInfo) {
        capturedInBeforeEach = testInfo
    }

    private fun assertSameIdentity(stackRoute: KotlinTestMeta, infoRoute: KotlinInfoBasedTestMeta) {
        assertEquals(stackRoute.testClassName(), infoRoute.testClassName(), "testClassName")
        assertEquals(stackRoute.testMethodName(), infoRoute.testMethodName(), "testMethodName")
        assertEquals(stackRoute.testClassPath, infoRoute.testClassPath, "testClassPath")
        assertEquals(stackRoute.approvedDirectory, infoRoute.approvedDirectory, "approvedDirectory")
        assertEquals(stackRoute.workingDirectory(), infoRoute.workingDirectory(), "workingDirectory")
    }

    @Test
    fun publicFunctionResolvesIdenticallyByBothRoutes(testInfo: TestInfo) {
        assertSameIdentity(KotlinTestMeta(), KotlinInfoBasedTestMeta(testInfo))
    }

    @Test
    internal fun internalFunctionResolvesIdenticallyByBothRoutes(testInfo: TestInfo) {
        assertSameIdentity(KotlinTestMeta(), KotlinInfoBasedTestMeta(testInfo))
    }

    @Test
    @DisplayName("a display name that is nothing like the function name")
    fun displayNameDoesNotAffectIdentity(testInfo: TestInfo) {
        val infoRoute = KotlinInfoBasedTestMeta(testInfo)

        assertSameIdentity(KotlinTestMeta(), infoRoute)
        assertEquals("displayNameDoesNotAffectIdentity", infoRoute.testMethodName(),
                "the function name, not the display name, identifies the test")
    }

    @Test
    fun testInfoCapturedInBeforeEachDescribesTheSameTest(testInfo: TestInfo) {
        assertEquals(
                KotlinInfoBasedTestMeta(testInfo).testMethodName(),
                KotlinInfoBasedTestMeta(capturedInBeforeEach!!).testMethodName()
        )
        assertSameIdentity(KotlinTestMeta(), KotlinInfoBasedTestMeta(capturedInBeforeEach!!))
    }

    /**
     * With the property unset, both routes must still resolve `src/test/kotlin`. This is the
     * regression that would relocate every Kotlin approved file in existence, so it is asserted
     * directly rather than left to the integration tests to notice.
     */
    @Test
    fun bothKotlinRoutesDefaultToSrcTestKotlin(testInfo: TestInfo) {
        val expected = java.nio.file.Paths.get("src/test/kotlin/com/github/karsaig/approvalcrest/kotlin")

        assertEquals(expected, KotlinTestMeta().testClassPath)
        assertEquals(expected, KotlinInfoBasedTestMeta(testInfo).testClassPath)
    }

    /**
     * `fileMatcherSourceRoot` is shared by every framework; only the fallback differs. Kotlin
     * projects whose tests live outside `src/test/kotlin` can therefore configure it like everyone
     * else.
     */
    @Test
    fun bothKotlinRoutesHonourTheSourceRootProperty(testInfo: TestInfo) {
        System.setProperty("fileMatcherSourceRoot", "src/it/kotlin")
        try {
            val expected = java.nio.file.Paths.get("src/it/kotlin/com/github/karsaig/approvalcrest/kotlin")

            assertEquals(expected, KotlinTestMeta().testClassPath,
                    "the stack route must honour the property")
            assertEquals(expected, KotlinInfoBasedTestMeta(testInfo).testClassPath,
                    "and the TestInfo route must honour it identically")
        } finally {
            System.clearProperty("fileMatcherSourceRoot")
        }
    }

    @Test
    fun bothKotlinRoutesHonourTheAlias(testInfo: TestInfo) {
        System.setProperty("fmSourceRoot", "src/it/kotlin")
        try {
            val expected = java.nio.file.Paths.get("src/it/kotlin/com/github/karsaig/approvalcrest/kotlin")

            assertEquals(expected, KotlinTestMeta().testClassPath)
            assertEquals(expected, KotlinInfoBasedTestMeta(testInfo).testClassPath)
        } finally {
            System.clearProperty("fmSourceRoot")
        }
    }

    @Nested
    inner class NestedTests {

        @Test
        fun nestedClassResolvesIdenticallyByBothRoutes(testInfo: TestInfo) {
            assertSameIdentity(KotlinTestMeta(), KotlinInfoBasedTestMeta(testInfo))
        }

        @Test
        fun nestedClassIsTheTestClass(testInfo: TestInfo) {
            assertEquals(NestedTests::class.java.name, KotlinInfoBasedTestMeta(testInfo).testClassName())
            assertEquals(NestedTests::class.java.name, KotlinTestMeta().testClassName())
        }
    }
}
